lighthouse-core/gather/gather-runner.js

```diff
@@ -187,17 +212,51 @@ class GatherRunner {
       const status = `Retrieving: ${gatherer.name}`;
       return chain.then(_ => {
         log.log('status', status);
-        return gatherer.afterPass(options, passData);
-      }).then(ret => {
+        const artifactPromise = Promise.resolve().then(_ => gatherer.afterPass(options, passData));
+        gathererResults[gatherer.name].push(artifactPromise);
+        return GatherRunner.recoverOrThrow(artifactPromise);
+      }).then(_ => {
         log.verbose('statusEnd', status);
-        return ret;
       });
     }, pass);
 
     // Resolve on tracing data using passName from config.
     return pass.then(_ => passData);
   }
 
+  /**
+   * Takes the results of each gatherer phase for each gatherer and uses the
+   * last produced value (that's not undefined) as the artifact for that
+   * gatherer. If a recoverable error was rejected from a gatherer phase,
+   * uses that error as the artifact instead.
+   * @param {!Object<!Array<!Promise<*>>} gathererResults
+   * @return {!Promise<!Artifacts>}
+   */
+  static collectArtifacts(gathererResults) {
+    const artifacts = {};
+
+    return Object.keys(gathererResults).reduce((chain, gathererName) => {
+      return chain.then(_ => {
+        const phaseResultsPromises = gathererResults[gathererName];
+        return Promise.all(phaseResultsPromises).then(phaseResults => {
+          // Take last defined pass result as artifact.
+          const definedResults = phaseResults.filter(element => element !== undefined);
+          const artifact = definedResults[definedResults.length - 1];
+          if (artifact === undefined) {
+            throw new Error(`${gathererName} failed to provide an artifact.`);
+          }
+          artifacts[gathererName] = artifact;
+        }, err => {
+          // To reach this point, all errors are recoverable, so return err to
+          // runner to handle turning it into an error audit.
+          artifacts[gathererName] = err;
+        });
+      });
+    }, Promise.resolve()).then(_ => {
```

Comment:
This is whole function is gnarly to grok, but I think the init value here can just be:

`}, Promise.resolve(artifacts));`